File: sidecars/windows/tests/Optimisarr.Sidecar.Core.Tests/ShutdownCountdownTests.cs
using Optimisarr.Sidecar.Core.Session;

namespace Optimisarr.Sidecar.Core.Tests;

public sealed class ShutdownCountdownTests
{
    private static readonly DateTimeOffset Now = new(2026, 9, 23, 12, 0, 0, TimeSpan.Zero);

    [Fact]
    public void Idle_worker_counts_down_only_after_a_confirmed_drained_heartbeat()
    {
        var plan = new ShutdownCountdown();
        plan.Arm();
        Assert.False(plan.Evaluate(Now, ready: false, activeJobs: 0));
        Assert.Equal("Waiting for the server to confirm draining", plan.Detail);
        Assert.False(plan.Evaluate(Now, ready: true, activeJobs: 0));
        Assert.Equal(60, plan.SecondsRemaining(Now));
        Assert.True(plan.Evaluate(Now.AddSeconds(60), ready: true, activeJobs: 0));
        Assert.True(plan.Begin());
        Assert.False(plan.Evaluate(Now.AddSeconds(61), ready: true, activeJobs: 0));
    }

    [Fact]
    public void Active_or_failed_job_must_finish_reporting_before_countdown()
    {
        var plan = new ShutdownCountdown();
        plan.Arm();
        Assert.False(plan.Evaluate(Now, ready: false, activeJobs: 1));
        Assert.Equal("Waiting for 1 job to finish verification, return, and acknowledgement", plan.Detail);
        Assert.False(plan.Evaluate(Now.AddSeconds(10), ready: true, activeJobs: 0));
        Assert.Equal(60, plan.SecondsRemaining(Now.AddSeconds(10)));
    }

    [Fact]
    public void Server_disconnect_resets_countdown_and_cancel_prevents_shutdown()
    {
        var plan = new ShutdownCountdown();
        plan.Arm();
        plan.Evaluate(Now, ready: true, activeJobs: 0);
        Assert.False(plan.Evaluate(Now.AddSeconds(30), ready: false, activeJobs: 0, serverUnreachable: true));
        Assert.Contains("Server unreachable", plan.Detail);
        Assert.False(plan.Evaluate(Now.AddSeconds(61), ready: true, activeJobs: 0));
        Assert.Equal(60, plan.SecondsRemaining(Now.AddSeconds(61)));
        plan.Cancel();
        Assert.False(plan.Evaluate(Now.AddSeconds(200), ready: true, activeJobs: 0));
    }

    [Fact]
    public void Permission_failure_is_visible_and_never_retried_automatically()
    {
        var plan = new ShutdownCountdown();
        plan.Arm();
        plan.Evaluate(Now, ready: true, activeJobs: 0);
        Assert.True(plan.Evaluate(Now.AddSeconds(60), ready: true, activeJobs: 0));
        Assert.True(plan.Begin());
        plan.Fail("Windows denied shutdown permission");
        Assert.Equal("Windows denied shutdown permission", plan.Detail);
        Assert.False(plan.Evaluate(Now.AddSeconds(120), ready: true, activeJobs: 0));
    }

    [Fact]
    public void Unacknowledged_result_blocks_shutdown_even_when_no_job_is_running()
    {
        var plan = new ShutdownCountdown();
        plan.Arm();
        Assert.False(plan.Evaluate(Now, ready: false, activeJobs: 0, unacknowledged: true));
        Assert.Contains("not acknowledged", plan.Detail);
        Assert.Null(plan.SecondsRemaining(Now));
    }

    [Fact]
    public void Cancel_during_final_server_check_still_prevents_power_off()
    {
        var plan = new ShutdownCountdown();
        plan.Arm();
        plan.Evaluate(Now, ready: true, activeJobs: 0);
        Assert.True(plan.Evaluate(Now.AddSeconds(60), ready: true, activeJobs: 0));
        Assert.True(plan.CanCancel);
        plan.Cancel();
        Assert.False(plan.Begin());
    }
}

File: sidecars/macos/Tests/SidecarCoreTests/ShutdownCountdownTests.swift
import Foundation
import Testing
@testable import SidecarCore

struct ShutdownCountdownTests {
    private let now = Date(timeIntervalSince1970: 1_800_000_000)

    @Test func idleNeedsConfirmedDrainThenCountsDown() {
        var plan = ShutdownCountdown()
        plan.arm()
        let unconfirmed = plan.evaluate(at: now, ready: false, activeJobs: 0)
        let started = plan.evaluate(at: now, ready: true, activeJobs: 0)
        #expect(!unconfirmed)
        #expect(!started)
        #expect(plan.secondsRemaining(at: now) == 60)
        let due = plan.evaluate(at: now.addingTimeInterval(60), ready: true, activeJobs: 0)
        let began = plan.begin()
        let repeated = plan.evaluate(at: now.addingTimeInterval(61), ready: true, activeJobs: 0)
        #expect(due)
        #expect(began)
        #expect(!repeated)
    }

    @Test func heldJobAndFailedJobAcknowledgementDelayCountdown() {
        var plan = ShutdownCountdown()
        plan.arm()
        let busy = plan.evaluate(at: now, ready: false, activeJobs: 1)
        #expect(!busy)
        #expect(plan.detail.contains("acknowledgement"))
        let released = plan.evaluate(at: now.addingTimeInterval(10), ready: true, activeJobs: 0)
        #expect(!released)
        #expect(plan.secondsRemaining(at: now.addingTimeInterval(10)) == 60)
    }

    @Test func disconnectResetsCountdownAndCancelStopsIt() {
        var plan = ShutdownCountdown()
        plan.arm()
        let started = plan.evaluate(at: now, ready: true, activeJobs: 0)
        let disconnected = plan.evaluate(at: now.addingTimeInterval(30), ready: false,
                                         activeJobs: 0, serverUnreachable: true)
        let disconnectedDetail = plan.detail
        let reconnected = plan.evaluate(at: now.addingTimeInterval(61), ready: true, activeJobs: 0)
        #expect(!started)
        #expect(!disconnected)
        #expect(disconnectedDetail.contains("Server unreachable"))
        #expect(!reconnected)
        #expect(plan.secondsRemaining(at: now.addingTimeInterval(61)) == 60)
        plan.cancel()
        let afterCancel = plan.evaluate(at: now.addingTimeInterval(200), ready: true, activeJobs: 0)
        #expect(!afterCancel)
    }

    @Test func deniedPermissionDoesNotRetry() {
        var plan = ShutdownCountdown()
        plan.arm()
        let started = plan.evaluate(at: now, ready: true, activeJobs: 0)
        let due = plan.evaluate(at: now.addingTimeInterval(60), ready: true, activeJobs: 0)
        let began = plan.begin()
        #expect(!started)
        #expect(due)
        #expect(began)
        plan.fail("macOS denied shutdown permission")
        #expect(plan.detail.contains("denied"))
        let repeated = plan.evaluate(at: now.addingTimeInterval(120), ready: true, activeJobs: 0)
        #expect(!repeated)
    }

    @Test func unacknowledgedResultBlocksShutdown() {
        var plan = ShutdownCountdown()
        plan.arm()
        let blocked = plan.evaluate(at: now, ready: false, activeJobs: 0, unconfirmed: true)
        #expect(!blocked)
        #expect(plan.detail.contains("not acknowledged"))
        #expect(plan.secondsRemaining(at: now) == nil)
    }

    @Test func cancelDuringFinalServerCheckPreventsShutdown() {
        var plan = ShutdownCountdown()
        plan.arm()
        _ = plan.evaluate(at: now, ready: true, activeJobs: 0)
        let due = plan.evaluate(at: now.addingTimeInterval(60), ready: true, activeJobs: 0)
        #expect(due)
        #expect(plan.canCancel)
        plan.cancel()
        let began = plan.begin()
        #expect(!began)
    }
}
